core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java

```diff
@@ -0,0 +1,213 @@
+  static final String PROXY_SCHEME = "https";
+  static final String URI_FORMAT = PROXY_SCHEME + "://%s";
+
+  /**
+   * A proxy selector that uses the global {@link ProxySelector#getDefault()} and
+   * {@link ProxyDetectorImpl.AuthenticationProvider} to detect proxy parameters.
+   */
+  public ProxyDetectorImpl() {
+    this(DEFAULT_PROXY_SELECTOR, DEFAULT_AUTHENTICATOR, System.getenv(GRPC_PROXY_ENV_VAR));
+  }
+
+  @VisibleForTesting
+  ProxyDetectorImpl(
+      Supplier<ProxySelector> proxySelector,
+      AuthenticationProvider authenticationProvider,
+      @Nullable String proxyEnvString) {
+    this.proxySelector = proxySelector;
+    this.authenticationProvider = authenticationProvider;
+    if (proxyEnvString != null) {
+      override = new ProxyParameters(overrideProxy(proxyEnvString), null, null);
+    } else {
+      override = null;
+    }
+  }
+
+  @Override
+  public ProxyParameters proxyFor(SocketAddress targetServerAddress) {
+    if (override != null) {
+      return override;
+    }
+    if (!(targetServerAddress instanceof InetSocketAddress)) {
+      return null;
+    }
+    InetSocketAddress targetInetAddr = (InetSocketAddress) targetServerAddress;
+    if (!targetInetAddr.isUnresolved()) {
+      /*
+       * If the address is already resolved, we can conclude that DnsNameResolver already called
+       * us earlier and we determined no proxy is needed.
+       */
+      return null;
+    }
+    return detectProxy(targetInetAddr);
+  }
+
+  private ProxyParameters detectProxy(InetSocketAddress targetAddr) {
+    String hostPort =
+        GrpcUtil.authorityFromHostAndPort(targetAddr.getHostName(), targetAddr.getPort());
+    URI uri;
+    try {
+      uri = new URI(String.format(URI_FORMAT, hostPort));
+    } catch (final URISyntaxException e) {
+      log.log(
+          Level.WARNING,
+          "Failed to construct URI for proxy lookup, proceeding without proxy: {0}",
+          hostPort);
+      return null;
+    }
+
+    List<Proxy> proxies = proxySelector.get().select(uri);
+    if (proxies.size() > 1) {
+      log.warning("More than 1 proxy detected, gRPC will select the first one");
+    }
+    Proxy proxy = proxies.get(0);
+
+    if (proxy.type() == Proxy.Type.DIRECT) {
+      return null;
+    }
+    InetSocketAddress proxyAddr = (InetSocketAddress) proxy.address();
+    PasswordAuthentication auth = authenticationProvider.requestPasswordAuthentication(
+        proxyAddr.getHostName(),
+        proxyAddr.getAddress(),
+        proxyAddr.getPort(),
+        PROXY_SCHEME,
+        "proxy authentication required by gRPC",
+        null);
+
+    if (auth == null) {
+      return new ProxyParameters(proxyAddr, null, null);
+    }
+
+    try {
+      //todo(spencerfang): users of ProxyParameters should clear the password when done
+      return new ProxyParameters(proxyAddr, auth.getUserName(), new String(auth.getPassword()));
+    } finally {
+      Arrays.fill(auth.getPassword(), '0');
```

Comment:
The javadoc for PasswordAuthentication recommends clearing the password after it is used. Even though gRPC doesn't treat passwords in a special way today, I think having one fewer place where passwords linger around is better.